services/src/logging.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2016 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "logging.h"
+
+#include <algorithm>
+#include <cstdio>
+#include <cstring>
+#include <cstdarg>
+#include "timer_hal.h"
+#include "service_debug.h"
+
+namespace {
+
+volatile log_message_callback_type log_msg_callback = 0;
+volatile log_write_callback_type log_write_callback = 0;
+volatile log_enabled_callback_type log_enabled_callback = 0;
+
+} // namespace
+
+void log_set_callbacks(log_message_callback_type log_msg, log_write_callback_type log_write,
+        log_enabled_callback_type log_enabled, void *reserved) {
+    log_msg_callback = log_msg;
+    log_write_callback = log_write;
+    log_enabled_callback = log_enabled;
+}
+
+void log_message_v(int level, const char *category, const char *file, int line, const char *func, void *reserved,
+        const char *fmt, va_list args) {
+    if (!log_write_callback && (!log_compat_callback || level < log_compat_level)) {
+        return;
+    }
+    const uint32_t time = HAL_Timer_Get_Milli_Seconds();
+    if (file) {
+        // Strip directory path
+        const char *p = strrchr(file, '/');
+        if (p) {
+            file = p + 1;
+        }
+    }
+    char buf[LOG_MAX_STRING_LENGTH];
+    if (log_msg_callback) {
+        const int n = vsnprintf(buf, sizeof(buf), fmt, args);
+        if (n > (int)sizeof(buf) - 1) {
+            buf[sizeof(buf) - 2] = '~';
+        }
+        log_msg_callback(buf, level, category, time, file, line, func, 0);
+    } else {
+        // Using compatibility callback
```

Comment:
It's used for compatibility with existent applications that were built against previous version of the firmware and use `SerialDebugOutput` for logging